src/identify_disks.c

```diff
@@ -0,0 +1,124 @@
+
+#include <dirent.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "debug.h"
+#include "identify_disks.h"
+#include "nvme.h"
+#include "util.h"
+
+/**
+ * Callback for scandir() to filter for NVMe namespaces.
+ */
+int is_nvme_namespace(const struct dirent *entry)
+{
+    unsigned int ctrl, nsid;
+    char p;
+
+    // Check for NVME controller name format: nvme<ctrl id>n<ns id>
+    return sscanf(entry->d_name, "nvme%un%u%c", &ctrl, &nsid, &p) == 2;
+}
+
+/**
+ * Enumerate namespaces for a controller.
+ */
+int enumerate_namespaces_for_controller(struct nvme_controller *ctrl)
+{
+    struct dirent **namelist;
+
+    int n = scandir(ctrl->sys_path, &namelist, is_nvme_namespace, alphasort);
+    if (n < 0)
+    {
+        fprintf(stderr, "failed scandir for %s: %m\n", ctrl->sys_path);
+        return 1;
+    }
+
+    DEBUG_PRINTF("found %d namespace(s) for controller=%s:\n", n, ctrl->name);
+    for (int i = 0; i < n; i++)
+    {
+        char namespace_path[MAX_PATH];
+        snprintf(namespace_path, sizeof(namespace_path), "/dev/%s", namelist[i]->d_name);
+
+        char *vs = nvme_identify_namespace_vs_for_namespace_device(namespace_path);
+        if (vs != NULL)
+        {
+            printf("%s: %s\n", namespace_path, vs);
+            free(vs);
+        }
+        free(namelist[i]);
+    }
+
+    free(namelist);
+    return 0;
+}
+
+/**
+ * Check if NVMe vendor in sysfs matches Microsoft's vendor ID.
+ */
+int is_microsoft_nvme_device(const char *device_name)
+{
+    char vendor_id_path[MAX_PATH];
+    snprintf(vendor_id_path, sizeof(vendor_id_path), "%s/%s/device/vendor", SYS_CLASS_NVME_PATH, device_name);
+
+    char *vendor_id_string = read_file_as_string(vendor_id_path);
```

Comment:
not that it matters much, but all the vendor udev files report 4096 as their length.

How about changing the utility to "read_hex_from_file" and doing:
```
FILE *f = fopen(argv[1], "r");
... // some error handling
long int x = 0;
int ret = fscanf(f, "%lx", &x);
if (ret != 1) {
   // return 0? return -1?
}
return x;
```

Reply:
I really like this idea but will do it in a follow up PR.

Reply:
actually, I removed util module and did it now! thanks!